proposals/handler.go

```diff
@@ -519,6 +523,87 @@ func (h *Handler) checkBallotSyntacticValidity(
 	return decoded, nil
 }
 
+func (h *Handler) getActiveSetWeight(ctx context.Context, id types.Hash32) (uint64, error) {
+	h.weightCalcLock.Lock()
+	totalWeight, exists := h.activeSets.Get(id)
+	if exists {
+		h.weightCalcLock.Unlock()
+		return totalWeight, nil
+	}
+
+	var ch chan uint64
+	chs, exists := h.pendingWeightCalc[id]
+	if exists {
+		// The calculation is running or the activeset is being fetched,
+		// subscribe.
+		// Avoid any blocking on the channel by making it buffered, also so that
+		// we don't have to wait on it in case the context is canceled
+		ch = make(chan uint64, 1)
+		h.pendingWeightCalc[id] = append(chs, ch)
+		h.weightCalcLock.Unlock()
+
+		// need to wait for the calculation which is already running to finish
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		case totalWeight, ok := <-ch:
+			if !ok {
+				// Channel closed, fetch / calculation failed.
+				// The actual error will be logged by the initiator of the
+				// initial fetch / calculation, let's not make an
+				// impression it happened multiple times and use a simpler
+				// message
+				return totalWeight, errors.New("error getting activeset weight")
+			}
+			return totalWeight, nil
+		}
+	}
+
+	// mark calculation as running
+	h.pendingWeightCalc[id] = nil
+	h.weightCalcLock.Unlock()
+
+	success := false
+	defer func() {
+		h.weightCalcLock.Lock()
+		// this is guaranteed not to block b/c each channel is buffered
+		for _, ch := range h.pendingWeightCalc[id] {
+			if success {
+				ch <- totalWeight
+			}
+			close(ch)
+		}
+		delete(h.pendingWeightCalc, id)
+		h.weightCalcLock.Unlock()
+	}()
+
+	if err := h.fetcher.GetActiveSet(ctx, id); err != nil {
+		return 0, err
+	}
+	set, err := activesets.Get(h.db, id)
+	if err != nil {
+		return 0, err
+	}
+	if len(set.Set) == 0 {
+		return 0, fmt.Errorf("%w: empty active set", pubsub.ErrValidationReject)
+	}
+
+	computed, used := h.atxsdata.WeightForSet(set.Epoch, set.Set)
+	for i := range used {
+		if !used[i] {
+			return 0, fmt.Errorf(
+				"missing atx %s in active set",
+				set.Set[i].ShortString(),
+			)
+		}
+	}
+	totalWeight = computed
+	h.activeSets.Add(id, totalWeight)
+	success = true // totalWeight will be sent to the subscribers
+
+	return totalWeight, nil
+}
```

Comment:
I'm a bit late to the discussion but I feel like this could have been done simpler using the `golang.org/x/sync/singleflight` package:

```go
func (h *Handler) getActiveSetWeight(ctx context.Context, id types.Hash32) (uint64, error) {
	result, err, _ := h.activeSetGroup.Do(id.String(), func() (any, error) {
		totalWeight, exists := h.activeSets.Get(id)
		if exists {
			return totalWeight, nil
		}

		if err := h.fetcher.GetActiveSet(ctx, id); err != nil {
			return uint64(0), err
		}
		set, err := activesets.Get(h.db, id)
		if err != nil {
			return uint64(0), err
		}
		if len(set.Set) == 0 {
			return uint64(0), fmt.Errorf("%w: empty active set", pubsub.ErrValidationReject)
		}

		computed, used := h.atxsdata.WeightForSet(set.Epoch, set.Set)
		for i := range used {
			if !used[i] {
				return uint64(0), fmt.Errorf("missing atx %s in active set", set.Set[i])
			}
		}
		h.activeSets.Add(id, computed)
		return computed, nil
	})
	if err != nil {
		h.activeSetGroup.Forget(id.String())
	}
	return result.(uint64), err
}
```

No mutex or channels needed 🙂